crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -632,24 +632,36 @@ impl CumulocityConverter {
         &mut self,
         message: &MqttMessage,
     ) -> Result<Vec<MqttMessage>, ConversionError> {
-        let operation = C8yOperation::from_json(message.payload.as_str()?)?;
-        let device_xid = operation.external_source.external_id;
-        let cmd_id = self.command_id.new_id_with_str(&operation.op_id);
+        // JSON over MQTT messages on c8y/devicecontrol/notifications can contain multiple operations in a single MQTT
+        // message, so split them
+        let operation_payloads = collect_c8y_messages(message.payload_str()?);
+
+        let mut output = vec![];
+        for operation_payload in operation_payloads {
+            let operation = C8yOperation::from_json(operation_payload.as_str())?;
+            let device_xid = operation.external_source.external_id;
+            let cmd_id = self.command_id.new_id_with_str(&operation.op_id);
+
+            if self.active_commands.contains(&cmd_id) {
+                info!("{cmd_id} is already addressed");
+                return Ok(vec![]);
+            }
 
-        if self.active_commands.contains(&cmd_id) {
-            info!("{cmd_id} is already addressed");
-            return Ok(vec![]);
-        }
+            // wrap operation payload in a dummy MqttMessage wrapper because the code below assumes 1 MQTT message = 1 operation
+            // TODO: refactor to avoid this intermediate step and extra copies
+            let operation_message = MqttMessage::new(&message.topic, operation_payload);
 
-        let result = self
-            .process_json_over_mqtt(
-                device_xid,
-                operation.op_id.clone(),
-                &operation.extras,
-                message,
-            )
-            .await;
-        let output = self.handle_c8y_operation_result(&result, Some(operation.op_id));
+            let result = self
+                .process_json_over_mqtt(
+                    device_xid,
+                    operation.op_id.clone(),
+                    &operation.extras,
+                    &operation_message,
+                )
+                .await;
+            let result = self.handle_c8y_operation_result(&result, Some(operation.op_id));
+            output.extend(result);
+        }
 
         Ok(output)
```

Comment:
My opinion for the future development.

What I don't like from the current way (due to the converter design) is, this return value Ok(output) is a vector of MQTT messages that will be published later at the same time. If 5 operations are included in one message from c8y, 5 of 504 messages will be sent together. But it's definitely out of scope to fix now.

Reply:
> What I don't like from the current way (due to the converter design) is, this return value Ok(output) is a vector of MQTT messages that will be published later at the same time.  If 5 operations are included in one message from c8y, 5 of 504 messages will be sent together.

What don't you like in the current design? That the value returned by this method is a vector of MQTT messages or that the five 504 messages are not packed into a single MQTT message? The later can be implemented using the former method signature.

Reply:
`504` changes an operation state to executing in c8y.

Let's say there are 3 operation (Op_A, Op_B, Op_C) are delivered at the same.
Ideally,
1. `Op_A` start proceeding -> Publish `504` for Op_A
2. `Op_B` start proceeding -> Publish `504` for Op_B
3. `Op_C` start proceeding -> Publish `504` for Op_C

However, with the current design, when all of `Op_A`, `Op_B`, and `Op_C` start proceeding, mapper publishes three `504` messages for these 3 operations. 

In short, what I don't like is the timing the `504` message is published by mapper. Not critical, but minor point.

Reply:
Though this doesn't occur for operations being processed by a workflow right (as the workflow execution is controlled by the tedge-agent, and currently only processes one type of operation at a time of the same type).